src/main/java/com/direwolf20/buildinggadgets/common/util/blocks/RegionSnapshot.java

```diff
@@ -0,0 +1,247 @@
+import net.minecraft.nbt.NBTTagCompound;
+import net.minecraft.nbt.NBTTagList;
+import net.minecraft.nbt.NBTUtil;
+import net.minecraft.tileentity.TileEntity;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.world.World;
+import net.minecraft.world.dimension.DimensionType;
+import net.minecraftforge.common.util.BlockSnapshot;
+import net.minecraftforge.common.util.Constants;
+import net.minecraftforge.fml.server.ServerLifecycleHooks;
+
+import java.util.*;
+import java.util.function.BiPredicate;
+
+public class RegionSnapshot {
+
+    private static final String DIMENSION = "dim";
+    private static final String BLOCK_FRAMES = "frames";
+    private static final String BLOCK_PALETTES = "palettes";
+    private static final String BLOCK_SNAPSHOTS = "block_snapshots";
+
+    private static NBTTagCompound serialize(NBTTagCompound tag, RegionSnapshot snapshot) {
+        tag.setInt(DIMENSION, snapshot.world.getDimension().getType().getId());
+        snapshot.region.serializeTo(tag);
+
+        // Palette serialization begin
+        NBTTagList palettes = new NBTTagList();
+        Object2IntMap<IBlockState> mapPalettes = new Object2IntOpenHashMap<>();
+        Set<IBlockState> recordedPalettes = new HashSet<>();
+
+        // "nothing should be here" block state
+        palettes.add(new NBTTagCompound());
+        // Palette serialization end
+
+        // Streaking: a way to record a number of repeating block states, using only 2 integers
+        // Storage frame: one or two integers indicating a sequence of block states
+        // Regular frame: a type of storage frame; 1 integer, representing the ID of the block state recorded
+        // Streak frame: a type of storage frame; 2 integers, where the first one is positive, representing the ID of the repeating block state; the second one is negative, representing the number of repeating block states
+
+        // Complete storage frames, will be serialized into a NBTTagList later
+        IntList frames = new IntArrayList();
+
+        // The repeating block state, if any
+        IBlockState lastStreak = null;
+        // Number of repeating block states
+        int streak = 0;
+
+        boolean firstIteration = true;
+        for (IBlockState state : snapshot.blockStates) {
+            // Palette serialization begin
+            if (!recordedPalettes.contains(state)) {
+                recordedPalettes.add(state);
+                // ID 0 is preoccupied for "nothing should be placed here", which is represented in 'null'
+                mapPalettes.put(state, mapPalettes.size() + 1);
+                palettes.add(NBTUtil.writeBlockState(state));
+            }
+            // Palette serialization ends
+
+            // Hence we record the storage frame after it is completed, the inital values of the counter would be used as a storage frame regardless, therefore we need to skip inorder for this algorithm to function correctly
+            if (firstIteration)
```

Comment:
Hence streak will be staying above 1 all the time, should we replace the firstIteration with a detection of streak == 0? Note that streak can be initialized as 0.

Reply:
Resolved cause it's a duplicate that somehow popped up